### DISCUSSION TITLE
How KEEP_REJECTS value is inferred

### DISCUSSION THREAD

**Discussion by zhaolianshuizls** [NVIDIA/cccl#1057] [CUB] [answered]
Cub code is very advanced for me, can you explain to me how the following code infers the template parameter `KEEP_REJECTS` value?

https://github.com/NVIDIA/cub/blob/b2e8bccb8c0cd15279974fe4b9b8d6fcd1842b57/cub/device/dispatch/dispatch_select_if.cuh#L554

        DispatchSelectIf dispatch(d_temp_storage,
                                    temp_storage_bytes,
                                    d_in,
                                    d_flags,
                                    d_selected_out,
                                    d_num_selected_out,
                                    select_op,
                                    equality_op,
                                    num_items,
                                    stream,
                                    ptx_version);

**Comment by gevtushenko** [comment, marked as answer]
Hello, @zhaolianshuizls! In this context, the `DispatchSelectIf` is an [injected class name](https://en.cppreference.com/w/cpp/language/injected-class-name), so the `KEEP_REJECTS` matches the one provided at the `DispatchSelectIf` implicit instantiation [here](https://github.com/NVIDIA/cccl/blob/020bfc433ab1a86e51e26c095ef191c9a52275df/cub/cub/device/device_select.cuh#L197-L213).

**Comment by zhaolianshuizls** [comment]
@senior-zero Thanks again for your help. Which cub repo should I look at? The cccl cub? Why cccl? What differs between this cub and cccl/cub?

**Reply by gevtushenko** [reply]
We are currently in the middle of transitioning to the CCCL repo, which aggregates CUDA C++ Core Libraries. The announcement with all the details is due to be published soon.
